Game/source/controllers/TowerController.hx

```diff
@@ -1,51 +1,148 @@
+        RenderBuffer.buffer.add(tower);
+
+        var level:Int = 0;
+        var newLayer:TowerLayer;
+        var xpos:Float;
+        var ypos:Float;
+        for(m in tower.rawMaterials)
+        {
+            xpos = tower.x+tower.origin.x;
+            ypos = tower.y+tower.origin.y-level*Constants.HEIGHT_OFFSET;
+
+            level++;
+            if(Type.getClass(m) == Foundation){
+                newLayer = new TowerLayer(xpos, ypos, AssetPaths.tower_layer__png, m.healthPoints);
+                tower.layers.add(newLayer);
+                RenderBuffer.buffer.add(newLayer);
+            }
+            else if(Type.getClass(m) == GunBase){
+                newLayer = new GunLayer(xpos, ypos, AssetPaths.gun_layer__png, m.healthPoints, m.attackPoints, level);
+                tower.layers.add(newLayer);
+                RenderBuffer.buffer.add(newLayer);
+            }
+        }
+        return tower;
+    }
+
+    public function takeDamage(obj:Attacker, tower:Tower):Void
+    {
+        if(obj.attackType==AttackType.Ground)
+        {
+            tower.layers.first().takeDamage(obj);
+            if(tower.layers.first().isDead){
+                destroyBottom(tower);
+            }
+        }
+        else if(obj.attackType==AttackType.Air)
+        {
+            tower.layers.last().takeDamage(obj);
+            if(tower.layers.last().isDead){
+                destroyTop(tower);
+            }
+        }
+
+        if(tower.layers.first() == null)
+        {
+            tower.isDead = true;
+        }
+    }
+
+    public function destroyBottom(tower:Tower):TowerLayer
+    {
+        var dequeued:TowerLayer = tower.layers.pop();
+        tower.rawMaterials.pop();
+
+        var level:Int = 0;
+        for(l in tower.layers)
+        {
+            level++;
+            l.changeLayerHeight(level);
+        }
+        return dequeued;
+    }
+
+    public function destroyTop(tower:Tower):TowerLayer
+    {
+        var popped:TowerLayer = tower.layers.last();
+        tower.rawMaterials.remove(tower.rawMaterials.last());
+
+        tower.layers.remove(tower.layers.last());
+        return popped;
+    }
+
+    public function shoot(tower:Tower, dist:Float, xTarget:Float, yTarget:Float):Void
+    {
+        for(gun in tower.layers)
+        {
+            if(Type.getClass(gun)==GunLayer)
+            {
+                //cast(gun, GunLayer);
+                if(cast(gun, GunLayer).shoot() && dist<=cast(gun, GunLayer).attackRange){
+                    //create bullet type specified by gun.ammoType
+                    var bullet:Projectile = new Projectile(gun.x+gun.origin.x, gun.y+gun.origin.y, xTarget, yTarget);
+                    RenderBuffer.buffer.add(bullet);
+                }
+            }
+        }
+    }
+
+    override public function update(tower:Tower):Void
+    {
+        super.update(tower);
 		
 		if (GameState.npcs[0] != null) {
-			trace(GameState.npcs[0].getX());
```

Comment:
I know this file's causing a merge conflict with the master branch. Please checkout my version as the old master branch code won't compile with my controller implementation.